### ISSUE TITLE
Cross Account AWS Lambda Layers

### ISSUE THREAD

**Issue by Arlington1985** [hashicorp/terraform-provider-aws#12728]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently, it's not possible to specify Lambda layers from a different AWS account. In cloud formation you can do it just with specifying _arn_ of the layer. It might be with the exact version or without version. But as if now, AWS has no capability to refer cross-account lambda layers without specifying version.

### New or Affected Resource(s)

* aws_lambda_layer_version

### Potential Terraform Configuration

```hcl
resource "aws_lambda_layer_version" "lambda_layer" {
  arn= "arn:aws:lambda:eu-west-1:123456789012:layer:layer_name:1"

  compatible_runtimes = ["nodejs8.10"]
}
```

### References

* https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-layerversion.html#cfn-lambda-layerversion-layername

**Comment by ewbankkit** [comment]
Similar:
* https://github.com/terraform-providers/terraform-provider-aws/issues/12689

@Arlington1985 Thanks for raising this. `arn` (and `layer_arn`) are both `Computed` attributes so cannot be set for resource creation.
Looking at the underlying Lambda API, [`layer_name`](https://docs.aws.amazon.com/lambda/latest/dg/API_PublishLayerVersion.html#API_PublishLayerVersion_RequestSyntax) could be an ARN (including a cross-account ARN) but I don't think the way that [`resourceAwsLambdaLayerVersionParseId`](https://github.com/terraform-providers/terraform-provider-aws/blob/37a73c0d43cfdf3cc91f57b632844206a1bfbf82/aws/resource_aws_lambda_layer_version.go#L247-L261) has been written would allow this.
I will investigate.

**Comment by Arlington1985** [comment]
Yes, you are right, instead of `resource` it should be `data`, but even this doesn't work.
But basically, I solved my problem just specifying the _arn_ in the `layers` parameter in 
`resource "aws_lambda_function"`. Still I think it should work also with the with referencing from `data`

**Comment by ewbankkit** [comment]
The [corresponding data source](https://www.terraform.io/docs/providers/aws/d/lambda_layer_version.html) uses the [`ListLayerVersions`](https://docs.aws.amazon.com/lambda/latest/dg/API_ListLayerVersions.html) API and it is unlikely that you would have permissions to call this on a 3rd-party layer.
e.g. for Datadog:

```console
$ aws --region us-west-2 lambda list-layer-versions --layer-name arn:aws:lambda:us-west-2:464622532012:layer:Datadog-Python37

An error occurred (AccessDeniedException) when calling the ListLayerVersions operation: User: arn:aws:iam::123456789012:user/kit is not authorized to perform: lambda:ListLayerVersions on resource: arn:aws:lambda:us-west-2:464622532012:layer:Datadog-Python37
```

**Comment by Arlington1985** [comment]
I think you might be specified version

**Comment by Arlington1985** [comment]
With 3rd party libs, it might be a problem, but if I am using just my own cross-account layer, then I will have option to give any permission

**Comment by JoeHitchenXtract** [comment]
The related issue which brought me here was trying to use the Scipy layer across multiple regions, because it seems both the region and the account number change in the ARN. I haven't yet found a good alternative to hardcoding the ARNs for the regions we want to use.

**Comment by fred-vogt** [comment]
AWS lambda layers support "resource level policies":
* https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html#configuration-layers-permissions

What is needed is similar to other resources that support attaching resource policies, such as [ECR repository resource policies](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecr_repository_policy).

**Comment by anwickes** [comment]
> 
> Seems there is already an enhancement request for this:
> 
> * [Add lambda_layer_version_permission resource #8595](https://github.com/terraform-providers/terraform-provider-aws/issues/8595)

My understanding of this issue is the inability to use the following data block to retrieve the latest version of a lambda layer that is shared from a different account.

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/lambda_layer_version.html

This would be handy where codebase A deploys a layer to account A and codebase B needs to retrieve the latest version of the layer to use with a lambda function in account B without having to open account A to see what version to use.

**Comment by aRobinson-R7** [comment]
We create lambda layers in a central account and make them accessible across the organization.
This way common code does not need to be uploaded to every account

**Comment by ZMarouani** [comment]
> Yes, you are right, instead of `resource` it should be `data`, but even this doesn't work. But basically, I solved my problem just specifying the _arn_ in the `layers` parameter in `resource "aws_lambda_function"`.
> 
> resource "aws_lambda_function" "lambda_function" {
>   ... 
>   layers = ["arn:aws:lambda:region:123456789012:layer:layer_name:version"]
>   ...
> Still I think it should work also with the with referencing from `data`

The problem with this solution is you will have to manually update the version each time if the layer only keeps the latest

**Comment by YakDriver** [comment]
As @ewbankkit noted in 2020, the real limitation is permissions. However, #46673 should now allow you to do anything your permissions allow. In other words, if you provide `layer_version_arn` for a cross account Lambda layer with or without a version number, and you have the proper permissions, the data source should work. There's nothing we can do about the permissions requirement as this is an API requirement. Hopefully this helps even if it doesn't solve that part of the problem.
